subprojects/parseq/src/main/java/com/linkedin/parseq/internal/ContextImpl.java

```diff
@@ -121,6 +120,40 @@ public void run(final Task<?>... tasks) {
     }
   }
 
+  private void submitToExecutorAsync(Task<?> task) {
+    // Spawn a subContext to drive the running of submitted task later
+    final ContextImpl subContext = new ContextImpl(_planContext, task, _parent, _predecessorTasks);
+    // Note:
+    // (1) We allow task async submission even current task is considered done(promise resolved),
```

Comment:
will this impact the criteria used in `ParSeqUnitTestHelper.runAndWait` and `runAndWaitForPlanComplete`? In another word, does this current plan need to wait for this task to finish to count the plan complete? since this will impact `Parseq PlanCompletionListener` triggering. 

Also, how is this API different from existing `side-effect` API? My take is that it differs from it in that it is not attached to a particular plan?

Reply:
The idea is to add a task to execution when the user executes the code in the lambda (at which time a plan is already started).

For side-effect to run, you still need to as the engine to trigger that, am I understanding this correctly?

Yes it will impact PlanCompletionListener (this task need to be done to trigger the planCompletion, so I think this is another difference?

Reply:
No, sideeffect is taken into consideration in `PlanCompletionListener`, only when side-effect task is finished, the listener is triggered for the plan.

Sorry, I still don't understand the motivation here: what do you mean "execution when the user executes the code in the lambda (at which time a plan is already started)."? What scenario cannot be done through current task API? I am not sure that I have a full grasp on the issue. I thought that you want to introduce a way to arbitrarily submit an orphan task to executor to run, but I don't know your expected outcome on how this async task will impact current plan or any plan it will be attached to?

Reply:
> "I thought that you want to introduce a way to arbitrarily submit an orphan task to executor to run" 
That is actually right.

> "What scenario cannot be done through current task API?"
Please see the issue description. I have created a task and scheduled to run it directly inside the lambda function. This is one goal. The task can also be created inside the lambda function of a `andThen` method (not shown in the description), so it can reuse parent task result, which is another goal. Similar thing will be complicated to do if to wrap it using `Task.async` and `Context.After().run()`. And as a matter of fact, I don't expect user to do touch `Context` at all (although some users still do). This API is context transparent to the API user.

> "how this async task will impact current plan or any plan it will be attached to"
Actually the parent/child relationship does not  seem to really matter too much in this regard, I only want the created task to be run as part of the plan.

Reply:
Why cannot they achieve the issue described using Task.withSideEffect https://javadoc.io/doc/com.linkedin.parseq/parseq/latest/com/linkedin/parseq/Task.html?  I still didn't get. Lamda or not is just a syntactic sugar on the API, I am still puzzled here.

Reply:
Syntax sugar, probably yes, but it will make a difference. I wish we could provide coroutine-alike experience to the end users, but we could not. So this `scheduleToRun` syntax sugar somewhat helps.

The main difference is that the created task returend by Task.withSideEffect will still need to be triggered by `engine.run()`, right? But this `scheduleToRun` is to schedule the task asynchronously and dynamically. To make an analogy, previously it was like run a computational graph by analyzing it statistically and then run (using Engine.run()) and now we are trying to make it dynamic. This api would make things otherwise need to be achieved by `sideEffect` easier. (similar to how the market favors PyTorch over Tensorflow 1.0? )

In order to prevent XY problem, the actual use case is for the universal client:
We are going to expose a Rest.Li implementation of `CompletionStage`, For example the user can put their logic inside the completationStage returned: Comparing the following two method
```java
CompletationStage method1() {
  CompletionStage album1 = Albums.get(1);
  return album1;
}

CompletationStage method2() {
  CompletionStage album1 = Albums.get(1);

  album1.thenCompose((album1) -> {
    log.info(
      AlbumDetail.get(album1.getSerialNumber())
      .toCompletableFuture().get()
    );


  Albums.get(2).thenCompose((albums2) -> {
    Singers.get(album1.getSinger()).thenCompose({
    (singer) -> updateDB(singerId,
      AlbumDetail.get(album2.getSerialNumber().toCompletableFuture.get()));
      })
    });

  });
  return album1;
}
```
When the user run either of the above code(either method1 or method2), the returned completionStage should already started resolving the values.
When the `thenCompose` becaming nested inside multiple layers, the difficulties of writing this using `sideEffect` will be bigger.

Reply:
also, Task.blocking is also serving the similar purpose.

Reply:
@mchen07 Hi Min, 
(1) First, how they are invoked are different:
the Task.blocking does serve the same purpose, but again you will have difficulty to trigger it in a lambda function other than manually use `engine.run` to invoke the task.

(2) Second, this API is for all ParSeq tasks. The purpose of blocking task and other parseq tasks are different. For mission critical blocking task, it will be best to run it in another executor service, so wrapping it in a blocking task and run in separate executorservice by calling `engine.run` made sense, but the user of this `scheduleToRun` will be calling it for tasks of different purpose.